js/core/options/index.js

```diff
@@ -0,0 +1,201 @@
+import createCallBack from '../utils/callbacks';
+import { isFunction, isObject, type } from '../utils/type';
+import { equalByValue } from '../utils/common';
+import { OptionManager } from './option_manager';
+import { clone } from '../utils/object';
+import { getFieldName, getParentName, convertRulesToOptions } from './utils';
+
+export class Options {
+    constructor(options, defaultOptions, optionsByReference, deprecatedOptions) {
+        this._deprecatedCallbacks = createCallBack({ syncStrategy: true });
+
+        this._default = defaultOptions;
+        this._deprecated = deprecatedOptions;
+
+        this._deprecatedNames = [];
+        this._initDeprecatedNames();
+
+        this._optionManager = new OptionManager(
+            options,
+            optionsByReference
+        );
+        this._optionManager.onRelevantNamesPrepared((options, name, value) => this._setRelevantNames(options, name, value));
+
+        this._rules = [];
+    }
+
+    set _initial(value) {
+        this._initialOptions = value;
+    }
+
+    get _initial() {
+        if(!this._initialOptions) {
+            const rulesOptions = this._getByRules(this.silent('defaultOptionsRules'));
+
+            this._initialOptions = this._default;
+            this._optionManager._setByReference(this._initialOptions, rulesOptions);
+        }
+
+        return this._initialOptions;
+    }
+
+    _initDeprecatedNames() {
+        for(const optionName in this._deprecated) {
+            this._deprecatedNames.push(optionName);
+        }
+    }
+
+    _getByRules(rules) {
+        rules = Array.isArray(rules) ? this._rules.concat(rules) : this._rules;
+
+        return convertRulesToOptions(rules);
+    }
+
+    _notifyDeprecated(option) {
+        const info = this._deprecated[option];
+
+        if(info) {
+            this._deprecatedCallbacks.fire(option, info);
+        }
+    }
+
+    _setRelevantNames(options, name, value) {
+        if(name) {
+            const normalizedName = this._normalizeName(name);
+
+            if(normalizedName && normalizedName !== name) {
+                this._setField(options, normalizedName, value);
+                this._clearField(options, name);
+            }
+        }
+    }
+
+    _setField(options, fullName, value) {
+        let fieldName = '';
+        let fieldObject = null;
+
+        do {
+            fieldName = fieldName ? `.${fieldName}` : '';
+            fieldName = getFieldName(fullName) + fieldName;
+            fullName = getParentName(fullName);
+            fieldObject = fullName ? this._optionManager.get(options, fullName, false) : options;
+        } while(!fieldObject);
+
+        fieldObject[fieldName] = value;
```

Comment:
Should we use `silent` method for these operations?

Reply:
We can't.
Silent doesn't trigger 'onOptionChanged' in component (but assigning do that). We caught incorrect behaviour.